File: Sources/TokenSpendie/Data/ClaudeProvider.swift
import Foundation

/// The `UsageProvider` for Claude Code. Composes the Keychain credential
/// reader and the `/api/oauth/usage` endpoint, retries once on a 401 by
/// re-reading the Keychain (Claude Code refreshes its token during normal
/// use), and converts the Claude-shaped `UsageSnapshot` into a generic
/// `ProviderSnapshot`.
struct ClaudeProvider: UsageProvider {
    let id: ProviderID = .claude
    let displayName: String = "Claude"

    private let credentials: CredentialStore
    private let endpoint: ClaudeUsageEndpoint

    init(credentials: CredentialStore = KeychainReader(),
         endpoint: ClaudeUsageEndpoint = EndpointUsageProvider()) {
        self.credentials = credentials
        self.endpoint = endpoint
    }

    func detectCredentials() -> Bool {
        credentials.credentialsExist()
    }

    func fetchUsage() async throws -> ProviderSnapshot {
        let creds = try credentials.loadCredentials()
        let usage: UsageSnapshot
        do {
            usage = try await endpoint.fetchUsage(accessToken: creds.accessToken)
        } catch ProviderError.unauthorized {
            // Re-read the Keychain once — Claude Code refreshes the token
            // during normal use — then retry.
            let refreshed = try credentials.loadCredentials()
            usage = try await endpoint.fetchUsage(accessToken: refreshed.accessToken)
        }
        return Self.convert(usage)
    }

    /// Pure `UsageSnapshot` → `ProviderSnapshot` mapping. The session window is
    /// the headline; `windows` is `[session, weekly, model-weeklies…]`.
    static func convert(_ usage: UsageSnapshot) -> ProviderSnapshot {
        let session = LabeledWindow(label: "Session", detail: "5-hour window",
                                    resetStyle: .countdown, window: usage.session)
        var windows = [session]
        windows.append(LabeledWindow(label: "Weekly", detail: "all models",
                                     resetStyle: .date, window: usage.weekly))
        for model in usage.modelWeeklies {
            windows.append(LabeledWindow(label: "Weekly · \(model.model)",
                                         detail: "\(model.model) only",
                                         resetStyle: .date, window: model.window))
        }
        return ProviderSnapshot(id: .claude, plan: nil, headline: session,
                                windows: windows, fetchedAt: usage.fetchedAt)
    }
}

File: windows/src/TokenSpendie.Windows/Data/ClaudeProvider.cs
using TokenSpendie.Windows.Models;

namespace TokenSpendie.Windows.Data;

public sealed class ClaudeProvider : IUsageProvider
{
    public ProviderID Id => ProviderID.Claude;
    public string DisplayName => "Claude";

    private readonly ICredentialReader _credentials;
    private readonly IClaudeUsageEndpoint _endpoint;

    public ClaudeProvider(ICredentialReader credentials, IClaudeUsageEndpoint endpoint)
    {
        _credentials = credentials;
        _endpoint = endpoint;
    }

    public bool DetectCredentials() => _credentials.CredentialsExist();

    public async Task<ProviderSnapshot> FetchUsageAsync(CancellationToken ct = default)
    {
        var creds = await _credentials.LoadCredentialsAsync(ct).ConfigureAwait(false);
        UsageSnapshot usage;
        try
        {
            usage = await _endpoint.FetchUsageAsync(creds.AccessToken, ct).ConfigureAwait(false);
        }
        catch (ProviderUnauthorizedException)
        {
            // Re-read credentials once — Claude Code refreshes the token during normal use.
            var refreshed = await _credentials.LoadCredentialsAsync(ct).ConfigureAwait(false);
            usage = await _endpoint.FetchUsageAsync(refreshed.AccessToken, ct).ConfigureAwait(false);
        }
        return Convert(usage);
    }

    /// <summary>Pure <see cref="UsageSnapshot"/> → <see cref="ProviderSnapshot"/> mapping.
    /// Session is the headline; <c>Windows</c> is <c>[session, weekly, model-weeklies…]</c>.</summary>
    public static ProviderSnapshot Convert(UsageSnapshot usage)
    {
        var session = new LabeledWindow("Session", "5-hour window",
            ResetStyle.Countdown, usage.Session);
        var windows = new List<LabeledWindow> { session };
        windows.Add(new LabeledWindow("Weekly", "all models", ResetStyle.Date, usage.Weekly));
        foreach (var model in usage.ModelWeeklies)
        {
            windows.Add(new LabeledWindow(
                $"Weekly · {model.Model}", $"{model.Model} only",
                ResetStyle.Date, model.Window));
        }
        return new ProviderSnapshot(
            Id: ProviderID.Claude, Plan: null,
            Headline: session, Windows: windows,
            FetchedAt: usage.FetchedAt);
    }
}
